File: 03_Stack/01_Valid_Parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        
        stack = []
        m = {
            '}': '{',
            ']': '[',
            ')': '(',
        }

        for x in s:
            if x == '(' or x == '{' or x == '[':
                stack.append(x)
            elif len(stack) == 0 or stack[-1] != m[x]:
                return False
            else:
                stack.pop()
        
        return len(stack) == 0

File: 03_Stack/01_Valid_Parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        Map<Character, Character> m = new HashMap<>();

        m.put('}', '{');
        m.put(']', '[');
        m.put(')', '(');

        for (char x : s.toCharArray()) {
            if (x == '{' || x == '[' || x == '(')
                st.push(x);
            else if (st.isEmpty() || st.peek() != m.get(x))
                return false;
            else
                st.pop();
        }

        return st.isEmpty();
    }
}